ir/src/main/scala/org/scalajs/core/ir/Infos.scala

```diff
@@ -64,4 +70,292 @@ object Infos {
+      this.kind = kind
+      this
+    }
+
+    def setIsExported(isExported: Boolean): this.type = {
+      this.isExported = isExported
+      this
+    }
+
+    def setSuperClass(superClass: Option[String]): this.type = {
+      this.superClass = superClass
+      this
+    }
+
+    def addInterface(interface: String): this.type = {
+      interfaces += interface
+      this
+    }
+
+    def addInterfaces(interfaces: TraversableOnce[String]): this.type = {
+      this.interfaces ++= interfaces
+      this
+    }
+
+    def addMethod(methodInfo: MethodInfo): this.type = {
+      methods += methodInfo
+      this
+    }
+
+    def result(): ClassInfo = {
+      ClassInfo(encodedName, isExported, kind, superClass,
+          interfaces.toList, methods.toList)
+    }
+  }
+
+  final class MethodInfoBuilder {
+    private var encodedName: String = ""
+    private var isStatic: Boolean = false
+    private var isAbstract: Boolean = false
+    private var isExported: Boolean = false
+
+    private val methodsCalled = mutable.Map.empty[String, mutable.Set[String]]
+    private val methodsCalledStatically = mutable.Map.empty[String, mutable.Set[String]]
+    private val staticMethodsCalled = mutable.Map.empty[String, mutable.Set[String]]
+    private val instantiatedClasses = mutable.Set.empty[String]
+    private val accessedModules = mutable.Set.empty[String]
+    private val accessedClassData = mutable.Set.empty[String]
+
+    def setEncodedName(encodedName: String): this.type = {
+      this.encodedName = encodedName
+      this
+    }
+
+    def setIsStatic(isStatic: Boolean): this.type = {
+      this.isStatic = isStatic
+      this
+    }
+
+    def setIsAbstract(isAbstract: Boolean): this.type = {
+      this.isAbstract = isAbstract
+      this
+    }
+
+    def setIsExported(isExported: Boolean): this.type = {
+      this.isExported = isExported
+      this
+    }
+
+    def addMethodCalled(receiverTpe: Type, method: String): this.type = {
+      receiverTpe match {
+        case ClassType(cls) => addMethodCalled(cls, method)
+        case AnyType        => addMethodCalled(ObjectClass, method)
+        case UndefType      => addMethodCalled(BoxedUnitClass, method)
+        case BooleanType    => addMethodCalled(BoxedBooleanClass, method)
+        case IntType        => addMethodCalled(BoxedIntegerClass, method)
+        case LongType       => addMethodCalled(BoxedLongClass, method)
+        case FloatType      => addMethodCalled(BoxedFloatClass, method)
+        case DoubleType     => addMethodCalled(BoxedDoubleClass, method)
+        case StringType     => addMethodCalled(StringClass, method)
+
+        case ArrayType(_, _) if method != "clone__O" =>
+          /* clone__O is overridden in the pseudo Array classes and is
+           * always kept anyway, because it is in scalajsenv.js.
+           * Other methods delegate to Object, which we can model with
+           * a static call to Object.method.
+           */
```

Comment:
The problem is that we don't have any reification of the `Array` class family in the info. We have them in types as `ArrayType`s, but not in the info.
Before this commit the situation was even wrong: if you created a small hello-world problem that only calls `notify()` on an `Array` (but on nothing else), the `Analyzer` would drop it.

It may be conceptually wrong, but at least it will never change. Whether we implement these things in `scalajsenv.js` or not, it will always be the case that all methods but `clone__O` dispatch to `Object`. And `Array.clone` is so weird that it can never be written in user-space.

If you have a better idea, I'm open. That's the best I could come up with.